### DISCUSSION TITLE
Order of loading ARX modules

### DISCUSSION THREAD

**Discussion by gswifort** [CEXT-Dan/PyRx#89] [General]
_Issue discussed in [#1](https://github.com/CEXT-Dan/PyRx/issues/1#issuecomment-2044200601) and [#86](https://github.com/CEXT-Dan/PyRx/discussions/86#discussioncomment-9822668)_.

In what order are ARX modules (and clones) loaded? From my observations, if we have modules installed globally (`%localappdata%\Programs\...`), `PyRx25.0.arx` is always loaded from the global directory. If there is no global installation, the module located in the same directory as `RxLoader25.0.arx` is taken - is this the intended effect? The `pyrx_onload.py` file is loaded from the same directory as `PyRx25.0.arx`.

```raw
K:\Programy\demos\__temp_pyrx_install\venv\Lib\site-packages\pyrx\PyRxBin\RxLoaderZ25.0.zrx loaded successfully.
Command: 
PyRxLoader::getPythonVenvPath K:\Programy\demos\__temp_pyrx_install\venv
PyRxLoader::thisModulePath K:\Programy\demos\__temp_pyrx_install\venv\Lib\site-packages\pyrx\PyRxBin\
PyRxLoader::getInstallPath C:\Users\gswi\AppData\Local\Programs\PyRx
PyRxLoader::getIniPath C:\Users\gswi\AppData\Local\Programs\PyRx\Bin\PyRx.INI
Loading PyRx from venv condition
K:\Programy\demos\__temp_pyrx_install\venv\Lib\site-packages\wx
PyRxLoader::PyRxLoader_loader Loading, C:\Users\gswi\AppData\Local\Programs\PyRx\Bin\PyRxZ25.0.Zrx
Finished PyRxLoader_loader
PyRx version <1.3.026.20240621> loaded:
Python Interpreter Loaded successfully!
```

**Comment by CEXT-Dan** [comment]
Yes, I think that’s correct. 
I don’t develop with the package installed, I have a visual studio debug config, copy PyRx25.0.arx, RxLoader25.0.arx, pyrx_onload.py, and relevant debugging files into the same folder, PYRXSTUBPATH is set to my repository.

Originally, I just stuck everything into the .bundle path, but I had to do something different for the clones 

I had also thought of installing everything in the same folder and have PackageContents.xml just point to the folder, but I think the user would have to add a trusted path. However, this might be worth revisiting from a python package installation perspective. Install everything into ..\Lib\site-packages, maybe get rid of the .ini file

**Reply by gswifort** [reply]
I got rid of the `.ini` file, global installation still takes priority. I think the local files should be prioritized. So if we have an active virtual environment, we load `%VIRTUAL_ENV%\Lib\site-packages\pyrx\PyRxBin\PyRx25.0.arx` (of course, this will only be possible if we change to installing the package via `pip`). This is important because of the versions. My software may work with a specific version of `PyRx` and sometimes may not work with a higher version.

**Reply by CEXT-Dan** [reply]
I don’t have a problem redoing the loader to work with pip
FYI, I’m on the road from the beginning in August, until about the middle of October.  I think it would be best for me to tackle this when I get back

**Reply by gswifort** [reply]
Hi, do you have any plans for a new loader in the near future? I'm working on python tools (you can see them [here](https://github.com/gswifort/PyRx/tree/py-lib/pyrx)) and I'd be happy if they were tested by a larger group.

**Reply by CEXT-Dan** [reply]
I want to, it can’t stay the way it is; however, I’m leaving again next Monday until the middle of October. I have task #90 and work on BricsCAD V25 changes. the list is piling up lol

**Reply by gswifort** [reply]
I understand, I'll try to do a lot of new cool stuff by then. Have a nice trip!

**Reply by gswifort** [reply]
>  [I’ll continue working on Bim wrappers, albeit with a lower priority](https://github.com/CEXT-Dan/PyRx/issues/90#issuecomment-2415295797)

Does this mean you will be working on a new loader in the near future? :smile:

**Reply by CEXT-Dan** [reply]
BricsCAD V25 comes out on 10/23. I have a boatload of support requests logged with Bricsys. I need to work through all the fixes and changes.

**Reply by gswifort** [reply]
Any plans on this topic?

**Reply by CEXT-Dan** [reply]
Do you mean “Order of loading ARX modules”?  or install from pip?
Install from pip will solve the order issue, correct?

**Reply by gswifort** [reply]
> Do you mean “Order of loading ARX modules”? or install from pip?

All :)

> Install from pip will solve the order issue, correct?

Probably yes.

**Comment by CEXT-Dan** [comment]
pyrx_onload.py is loaded from AutoCAD’s findfile search order
https://help.autodesk.com/view/OARX/2023/ENU/?guid=OARX-RefGuide-acedFindFile_ACHAR___AcString__

My reasoning is people might want to have a separate pyrx_onload routines for a particular drawing or project. 
I know this contradicts my earlier thought of having multiple files floating around would cause trouble.
I still think it will, but I think the flexibility is cool..
